Add CFG2 cross matching notebook

jleagle94 opened the pull request:
Adds a new directory for a Fornax notebook and files demonstrating high-energy Galactic source studies with Chandra, Fermi, GAIA and 2MASS. 

Closes #491

jleagle94 committed:
Review and add new fornax notebook and files

bsipocz reviewed (commented):
I had a quick run though it, haven't tried to run the code or render the notebook.

Overall it feels that there could and should be way more narrative, especially at the last third of the notebook

trjaffe commented on all_sky_galactic_chandra_fermi_gaia_2mass/all_sky_galactic_chandra_fermi_gaia_2mass.md:
````diff
@@ -0,0 +1,967 @@
+
+## Step 2: Use HEASARC VO Service to access Chandra observation metdata
+
+```python
+heasarc = vo.dal.TAPService("https://heasarc.gsfc.nasa.gov/xamin/vo/tap")
+```
+
+## Step 3: Run a ADQL query to get all observation positions
+
+```python
+query = """
+SELECT TOP 99999999 obsid, ra, dec, exposure, detector, time, name
+FROM chanmaster
+WHERE ra IS NOT NULL AND dec IS NOT NULL
+"""
+
+
+results = heasarc.search(query)
+table = results.to_table()
+table[:5]
````
I asked her to use PyVO where possible. If you already know that the data are at HEASARC, and if your query can be done by astroquery, then it is indeed simpler.  But note that this is a call to fetch the entire chanmaster catalog.  I'm not sure astroquery.heasarc can do that.  If you use query_region() and give a full-sky search radius, it'll slow the query down significantly.  

Is there an astroquery-y way to fetch an entire catalog that's served out of a database (as opposed to a file on disk)?

bsipocz replied:
yes, a `query_tap` via astroquery is possible, one can use that for custom ADQL as opposed to the wrapper/convenience of a spatial TAP query via `query_region`

bsipocz replied:
> Is there an astroquery-y way to fetch an entire catalog that's served out of a database (as opposed to a file on disk)?

I don't understand this, astroquery doesn't require file on disk for anything (well, it's some of the super small modules that do that typically when it's just a something dumped on a website, but it's rare and the exception). It does queries whenever there are an actual server on the other end.

trjaffe commented on all_sky_galactic_chandra_fermi_gaia_2mass/all_sky_galactic_chandra_fermi_gaia_2mass.md:
````diff
@@ -0,0 +1,967 @@
+
+It is interesting to note the extragalactic source classes falling within the Galactic plane. Can we cross match the table we just created with IRSA and MAST to gather IR and optical counterparts? IRSA includes 2MASS (IR) and MAST has GAIA (optical). 
+
+```python
+gaia_dr3 = vo.dal.TAPService("https://gea.esac.esa.int/tap-server/tap")
+mass = vo.dal.TAPService("https://irsa.ipac.caltech.edu/TAP")
+```
````
This is a nice example where we can demonstrate the end-to-end PyVO way of doing things, particularly to demonstrate the data discovery so that users can modify it for a different dataset.  (It also depends on whether it will be used later with a constraint that is not just spatial.  You cannot do generic SQL queries like that in most astroquery modules, AFAIK.)  

We (HEASARCers) think we may present both possibilities, hiding PyVo maybe under a "for a more generic solution, see here" or something.   But do we need to do that in every single notebook?  That would be a pain.  But again, what does the user do when they don't know which astroquery module serves the data they might need?  Is there any sort of astroquery registry?  Searching astroquery docs for a module that serves the data you want?  That could be cool, and an alternative to the VO Registry, which has issues.  

To be continued.

bsipocz replied:
you can do generic adql at most of the big astroquery modules, the rest are either not VO complient, or not maintained by the archives, or using vo is just not their preference. For IRSA/2MASS I think it's a reasonable ask to use our astroquery module. --> and here there is no data discovery, the users are expected not just to know the name of the archives but to know the exact URLs, which is not a reasonable expectation.

bsipocz replied:
for data access I would think it would be great to get Jessica involved in the brainstorming, too. I mean we did add a lot of data access already (some of which may have room for improvements), and just overall to coordinate on the exact topic as some of the resources may exist already

bsipocz replied:
> We can coordinate collaboration over on [HEASARC/tutorials#3](https://github.com/HEASARC/tutorials/issues/3).

Not us, that's a private repo.

> Of course data access notebooks do exist, but from what I can tell, none that are Fornax-specific

That's the point, data access doesn't have to have anything fornax specific in them, we indeed indeed have some cloud access notebooks, but given that all the data is public it should work in any other jupyterlab, so I really don't get what you mean Fornax-specific data access.
(Very early on we had some goofys mounts, but those are long gone and IMO are not planned to add back)

jleagle94 replied:
It used to be in fornax-demo-notebooks, but it was moved... 

Fornax-specific data access means demonstrating how to access and utilize all archival data. I personally would like to see such a notebook to expand my research interests to utilize IRSA and MAST. This is the point of Fornax, right? As a high-energy person myself, I am unfamiliar with IRSA and MAST and how they have become accustomed to storing, accessing, and using their data, and would like to learn how to. I feel like this is a shared thought for many of us.

bsipocz replied:
yes, but there is zero fornax specific of that data access, you will be using standard tools and protocols.

Fornax specific would be to bake in local paths and code that is only available on fornax. Fortunately both the libraries and the data are public -- there are multiple reasons to do all of this within fornax, but none of those reasons are that there are some fornax specific data access for otherwise non-public data.

jleagle94 replied:
Thanks for the additional context, @troyraen. I know that each archive already has some data access notebooks, but as @trjaffe has pointed out, there are differences in how each archive stores and accesses data in the cloud. I think it would be useful to have a multi-archive notebook that can touch on this a bit, and also inform users of the various tools and methods that are at their disposal. It would be a bit more comprehensive than what I have seen in single archive notebooks. 

I have a current data access notebook that tries to do this that is under review with HEASARC right now that might make its way here at some point. The same version is available [here](https://github.com/jleagle94/high-energy_notebook_tutorials/blob/main/fornax_notebooks/data_access/data_access.md). The point of the notebooks I am developing is to exactly emphasize one of your points about Fornax: that data access changes depending on the goal of the user and to demonstrate what is available and when it might be useful. 

I would like to know more about what use cases are well suited for a Fornax environment so I can make sure the ones I develop take advantage of this.

troyraen commented on all_sky_galactic_chandra_fermi_gaia_2mass/all_sky_galactic_chandra_fermi_gaia_2mass.md:
```diff
@@ -0,0 +1,967 @@
+* astropy
+* pyvo
+* matplotlib
+* healpy
```
healpy is not actually being used in the notebook, so this can be removed from the dependencies and imports. If you do need something like this in the future, I recommend [hpgeom](https://hpgeom.readthedocs.io/). It's much lighter weight, won't cause dependency issues here, and further it's much easier to install cross-platform (Linux, Windows, Mac). The only functionality healpy has that hpgeom doesn't is plotting. So if that's what you need, you may want to consider looking into the dependency issues and/or using something like matplotlib.

jleagle94 replied:
Yes sorry, I forgot to check that everything listed is actually being used. Thanks for catching that.

troyraen commented on all_sky_galactic_chandra_fermi_gaia_2mass/README.md:
This repo now has a `crossmatch` directory. Add this there if you feel that's an appropriate descriptor.

jleagle94 replied:
I will change it to CFG2_cross-matching.

bsipocz replied:
fully agree, my comment above was about naming the notebook, but now I see it was easy to misinterpret (or more precisely, to not read my mind...)

jleagle94 commented:
Thanks for your comments @bsipocz. I am working on implementing them over the next week or so. I will explore astroquery to see if it simplifies the code.

troyraen reviewed (commented):
Thanks @jleagle94, it's fun to see some high-energy data! My background isn't in this area but it made me interested to learn more.

I ran the notebook on the Fornax Science Console and it ran without error. Reading through it, I was mostly able to follow _what_ you're doing but had a hard time following the why. I would have loved some intro text before jumping in and a quick note in many places to help me learn about the data. For this PR, I would say please add a title and an introduction giving a little bit of background about the datasets and what can be learned by putting them together. Also fix the requirements file so it can be used reliably to run the notebook.

I left other comments below with an eye towards the review checklists (https://github.com/nasa-fornax/fornax-demo-notebooks/blob/main/notebook_review_checklists.md) but historically we don't expect that everything is addressed in the first PR. The existing workflow for this repo is a 2-step process. First step is into the repo and second is into the rendered docs. We review using the checklists more explicitly before finishing step two. There's a notebook template in prep in https://github.com/nasa-fornax/fornax-demo-notebooks/pull/473. Feel free to comment. Of course, all of these process and related docs can and should continue to be iterated on. Just letting you know where we're at up til this point.

jleagle94 commented on all_sky_galactic_chandra_fermi_gaia_2mass/all_sky_galactic_chandra_fermi_gaia_2mass.md:
````diff
@@ -0,0 +1,967 @@
+plt.show()
+```
+
+The code below plots the matched table along the Galactic plane again, but using the Astropy WCS package so we can display only the Galactic plane.
````
Well, it offers two ways to plot the data, and zooming in on the part of the sky that we are actually interested in can be more informative than an all-sky map, but to do a zoomed in sky map I cannot use the aitoff projection. I don't know if it is more than just a little coding exercise 😀

jleagle94 commented:
Thanks for the comments and background info @troyraen. I will definitely improve the narration throughout the notebook, related to many of the comments of others.

trjaffe commented on all_sky_galactic_chandra_fermi_gaia_2mass/all_sky_galactic_chandra_fermi_gaia_2mass.md:
````diff
@@ -0,0 +1,967 @@
+    print(col.name, col.datatype, "-", col.description, col.unit)
+```
+
+Unfortunately adding DISTANCE() breaks the mass ADQL search. <a href="https://irsa.ipac.caltech.edu/docs/program_interface/TAP.html">IRSA TAP docs</a> 
````
```
twomass_query = f"""
    SELECT TOP 3 designation, ra, dec, glon, glat, DISTANCE(POINT('ICRS', ra,dec),POINT('ICRS', {f_coord.ra.deg}, {f_coord.dec.deg})) AS dist
    FROM fp_psc
    WHERE 1=CONTAINS(
        POINT('ICRS', ra, dec),
        CIRCLE('ICRS', {f_coord.ra.deg}, {f_coord.dec.deg}, {radius})
    )
    ORDER BY dist ASC
    """
vo.dal.TAPService("https://irsa.ipac.caltech.edu/TAP").search(twomass_query).to_table()
```

Works for me.  What's the issue?

jleagle94 committed:
Merge branch 'nasa-fornax:main' into add-notebook

jleagle94 commented on all_sky_galactic_chandra_fermi_gaia_2mass/all_sky_galactic_chandra_fermi_gaia_2mass.md:
```diff
@@ -0,0 +1,967 @@
+
+import requests
+
+#2MASS query
```
What version of pyvo are you using?

jleagle94 replied:
I apologize, my intention is not being conveyed here well. In the initial tests I have run so far, an astroquery search within HEASARC data took nearly twice as long (~30 s) as the identical command with pyvo (~15 s). This is only preliminary and may not be significant for other queries.

I haven’t fully explored other differences yet, but since this part of the code already takes several minutes, I need more time to properly evaluate both before integrating astroquery. I would appreciate additional time to explore these tools and determine the best way to implement them here.

troyraen commented on all_sky_galactic_chandra_fermi_gaia_2mass/all_sky_galactic_chandra_fermi_gaia_2mass.md:
```diff
@@ -0,0 +1,967 @@
+    """
+    gaia_results = gaia_dr3.run_async(query).to_table()
+    
+    #2MASS query
+    twomass_query = f"""
+    SELECT TOP 3 designation, ra, dec, glon, glat
+    FROM fp_psc
+    WHERE 1=CONTAINS(
+        POINT('ICRS', ra, dec),
+        CIRCLE('ICRS', {f_coord.ra.deg}, {f_coord.dec.deg}, {radius})
+    )
+    """
+    twomass_results = mass.run_async(twomass_query).to_table()
```
I just checked our docs and I believe it's possible cross match all sources at the same time by doing a table upload. That should be faster and require less code so will be much better than looping over the coordinates one by one. See https://irsa.ipac.caltech.edu/docs/program_interface/TAP.html#upload (you probably want "8. Upload local file").

I also see `DISTANCE()` there, so I think that should work. It looks like that worked for Tess above, so maybe you figured that out already.

jleagle94 replied:
You are absolutely right. Abdu pointed this out to me about a week ago and I have updated the query to use a table upload for GAIA. However, I was having issues using it for 2MASS, so for now it is the only query still using a single query per source. If I recall, it was the JOIN() function needed to include the table upload that was not being recognized.

Though for sure the code is much faster this way! Would be great to include it for 2MASS too if I can. 

I have explored the DISTANCE() issue and I narrowed it down to the following. It works until a coordinate becomes negative. Then it breaks. I have informed the IRSA help desk of this and for now the query remains to exclude a DISTANCE() function. 

The query is definitely looking even more complicated though - so I am hoping an astroquery version will look nicer. It is the next step I will work on.

jleagle94 committed:
Merge branch 'nasa-fornax:main' into add-notebook

jleagle94 committed:
Rename all_sky_galactic_chandra_fermi_gaia_2mass/README.md to crossmatch/cxc-fermi-gaia-2mass_cross-match/README.md

jleagle94 committed:
Update README.md

jleagle94 committed:
Rename all_sky_galactic_chandra_fermi_gaia_2mass/all_sky_galactic_chandra_fermi_gaia_2mass.md to crossmatch/cxc-fermi-gaia-2mass_cross-match/cxc-fermi-gaia-2mass_cross-match.md

jleagle94 committed:
Delete all_sky_galactic_chandra_fermi_gaia_2mass directory

jleagle94 committed:
requirements file

jleagle94 committed:
Updated notebook file

Following notebook review comments. Need to implement astroquery in 2MASS query at the end of the notebook. Look for it in the next version.

jleagle94 committed:
Update index.md

jleagle94 committed:
Final versions

jleagle94 commented:
I have implemented most of everyone's comments/suggestions following the discussions here. Latest version is up. I ran into a new issue running it on Fornax though, which I think is coming from Python 3.12 dependency clashes between matplotlib and astropy. I don't have time to work through this, so I instead made a Python 3.11 environment and ran the notebook there successfully. The version in shared-storage/ on Fornax documents those steps needed for now.

troyraen committed:
Merge branch 'main' into add-notebook
